org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/AbstractProjectTemplate.java

```diff
@@ -0,0 +1,122 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.wizard.template;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.log4j.Logger;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+
+/**
+ * A template definition for a project, used by the new project wizard. Defines the UI (label, description, icon, variables) on how to
+ * present the template to the user and a list or {@link ProjectFile} elements to be generated by the wizard when the template is selected.
+ * The {@link ProjectFile}'s can contain references to {@link ProjectVariable}'s to allow the user to configure the generated code.
+ * 
+ * Instances of this class are generated from classes annotated with {@link ProjectTemplate}.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ * @since 2.14
+ */
+public abstract class AbstractProjectTemplate {
+
+	private static final Logger logger = Logger.getLogger(AbstractProjectTemplate.class);
+
+	private List<ProjectVariable> variables;
+
+	protected final StringProjectVariable text(String name, String defaultValue, String description) {
+		return new StringProjectVariable(name, defaultValue, description);
+	}
+
+	protected final BooleanProjectVariable check(String name, boolean defaultValue, String description) {
+		return new BooleanProjectVariable(name, defaultValue, description);
+	}
+
+	protected final StringSelectionProjectVariable combo(String name, String[] possibleValues, String description) {
+		return new StringSelectionProjectVariable(name, possibleValues, description);
+	}
+
+	/**
+	 * @return The label read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getLabel() {
+		return getProjectTemplateAnnotation().label();
+	}
+
+	/**
+	 * @return The icon read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getIcon() {
+		return getProjectTemplateAnnotation().icon();
+	}
+
+	/**
+	 * @return The description read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getDescription() {
+		return getProjectTemplateAnnotation().description();
+	}
+
+	/**
+	 * @return The variables read reflectively from this class.
+	 */
+	public final List<ProjectVariable> getVariables() {
+		if (variables == null) {
+			// lazy initialization because initialization in constructor would initialize lists with null's because field initialization
+			// comes after constructor initialization of super class
+			variables = new ArrayList<>();
+			Class<? extends Object> projectTemplateClass = getClass();
+			for (Field field : projectTemplateClass.getDeclaredFields()) {
```

Comment:
Would it make sense to generate the setters in a way that the field values are put into a list without using reflection?

Reply:
Yes, you are ritght. The reflection is an left over from an early implementation and can now be replaced with type safe code. Will do so.